[reverted] WAS: refactor(options): set option value for non-current context directly

famiu opened the pull request:
Problem: Currently, we use `switch_option_context` to temporarily switch the current option context before setting an option for a different buffer / window. This is not ideal because we already support getting and setting option values for non-current contexts in the underlying implementation.

Solution: Set option value for non-current context by passing the context directly to the lower level functions. Also introduce a new `OptCtx` struct to store option context information, this will scale much better if we add more option scopes and other context information in the future.

famiu committed:
refactor(options): set option value for non-current context directly

Problem: Currently, we use `switch_option_context` to temporarily switch the current option context before setting an option for a different buffer / window. This is not ideal because we already support getting and setting option values for non-current contexts in the underlying implementation.

Solution: Set option value for non-current context by passing the context directly to the lower level functions. Also introduce a new `OptCtx` struct to store option context information, this will scale much better if we add more option scopes and other context information in the future.

zeertzjq commented:
Won't this break autocommands?

famiu commented:
> Won't this break `OptionSet` autocommands?

I did not think about that... I wonder if there is any way we could avoid that? Maybe by putting additional data on the autocmd?

zeertzjq commented:
There may be problems with ftplugins as well

famiu commented:
@lewis6991 would you happen to have any advice regarding this?

zeertzjq commented:
Of course it still is. Checking for `.buf` (and possible calling `nvim_buf_call`) will make an autocommand much more verbose (especially if it was a one line Vimscript autocommand).

lewis6991 commented:
> > Won't this break `OptionSet` autocommands?
> 
> Is this still the case now that `ctx.buf` is passed as buffer to `apply_autocmds`?

Curbuf needs to be set otherwise it is a breaking change.

famiu commented:
> > > Won't this break `OptionSet` autocommands?
> > 
> > 
> > Is this still the case now that `ctx.buf` is passed as buffer to `apply_autocmds`?
> 
> Curbuf needs to be set otherwise it is a breaking change.

That is now done (using `aucmd_prepbuf`)

zeertzjq commented on src/nvim/option_defs.h:
```diff
@@ -81,6 +81,12 @@ typedef struct {
   OptValData data;
 } OptVal;
 
+// Context that an option is being set for.
```
```suggestion
/// Context that an option is being set for.
```

zeertzjq reviewed (commented):
`do_syntax_autocmd()` and `do_spelllang_source()` needs the context switch as well.

zeertzjq commented:
> this will scale much better if we add more option scopes and other context information in the future.

Why?

famiu commented:
> > this will scale much better if we add more option scopes and other context information in the future.
> 
> Why? I can't understand how this can happen.

As per the proposal in #29314, a tab scope may be added in the future, it's better to pass a single struct over 3 function parameters

famiu commented:
> `do_syntax_autocmd()` and `do_spelllang_source()` needs the context switch as well.

Why does `do_spelllang_source` need it? It already seems to take a window param? Does it trigger any autocmds?

famiu commented:
Just noticed that this does not handle TTY options, marking as draft until that's resolved

justinmk commented on src/nvim/option.c:
```diff
@@ -4572,299 +4447,330 @@ void *get_varp_from(vimoption_T *p, buf_T *buf, win_T *win)
-static inline OptIndex get_opt_idx(vimoption_T *opt)
-  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_PURE
+/// Get the current context for options.
+OptCtx option_get_cur_ctx(void)
```
```suggestion
OptCtx option_ctx(void)
```

justinmk commented on src/nvim/option.c:
```diff
@@ -1874,14 +1875,18 @@ static void apply_optionset_autocmd(OptIndex opt_idx, int opt_flags, OptVal oldv
   }
-  apply_autocmds(EVENT_OPTIONSET, options[opt_idx].fullname, NULL, false, NULL);
+
+  WITH_AUCMD_CONTEXT(ctx.buf, {
+    apply_autocmds(EVENT_OPTIONSET, options[opt_idx].fullname, NULL, false, NULL);
+  });
```
is this a behavior change and if so is there a way to cover it with a test?

famiu replied:
It's not really a behavior change. Previously this context change happened throughout the entirety of `set_option`, now the context change has been narrowed down to the specific parts that need it.

justinmk commented on src/nvim/option_defs.h:
```diff
@@ -81,6 +81,12 @@ typedef struct {
   OptValData data;
 } OptVal;
 
+/// Context that an option is being set for.
+typedef struct {
+  win_T *win;
+  buf_T *buf;
+} OptCtx;
```
this is probably generalizable / applicable to all kinds of stuff, not only options. similar to `vim._with`. see also `context.c`

famiu commented:
Been trying to make an `aucmd_prepbuf_win` function that also takes a window parameter, some tests are failing, so working on debugging that rn

famiu committed:
refactor(autocmd): allow specifying window for autocmd context

Problem: Currently we can only specify a buffer to use for an autocmd context through `aucmd_prepbuf()`, which finds a window that uses that buffer in the current tabpage, or creates an autocmd window. This means it's not possible to actually specify a window to use for an autocmd.

Solution: Add an `aucmd_prepbuf_win()` function which also takes a window as a parameter and uses it for the autocmd. If the window is not provided, then it behaves similarly to `aucmd_prepbuf()`

justinmk commented:
Is this waiting for something? Did anyone have objections?

famiu commented:
Trying to daily drive this PR, and seems like some peculiar and hard-to-reproduce bugs are popping up when I'm using plugins. Wonder if we should revert it.

zeertzjq commented:
This PR seems to also cause `Test_vimgrep()` to produce several different failures frequently on Windows:
```
Found errors in Test_vimgrep_autocmd():
command line..script D:/a/neovim/neovim/test/old/testdir/runtest.vim[581]..function RunTheTest[55]..Test_vimgrep_autocmd line 22: command did not fail: vimgrep stars Xtest*.txt
```
```
Found errors in Test_vimgrep():
command line..script D:/a/neovim/neovim/test/old/testdir/runtest.vim[581]..function RunTheTest[55]..Test_vimgrep[1]..XvimgrepTests line 61: Expected 1 but got 2
command line..script D:/a/neovim/neovim/test/old/testdir/runtest.vim[581]..function RunTheTest[55]..Test_vimgrep[2]..XvimgrepTests line 61: Expected 1 but got 3
```
With this PR: https://github.com/zeertzjq/neovim/actions/runs/12683272344
Without this PR: https://github.com/zeertzjq/neovim/actions/runs/12683334094
